README.md
# htop-2.1.0 (Ubuntu 18.04 - arm64 H5)

htop with CPU Temp and CPU Freq for arm64 H5 boards (mainline kernel)

[![htop](https://github.com/avafinger/htop-2.1.0/raw/master/img/htop.png)]



debian/changelog
htop (2.1.0-3) unstable; urgency=medium

  * Fix case of dyscalculia in changelog entry for 2.1.0-1
  * Update Daniel Lange's email address
  * Add debian/clean to remove autogenerated headers
  * Refresh debian/patches/601-openvz-new-ctid-vpid.patch
  * Fix issue with small terminals (Closes: #889693)
  * Fix preservation of LDFLAGS value during configure script
  * Handle unexpected values for character passed to isalnum

 -- Graham Inggs <ginggs@debian.org>  Sat, 17 Feb 2018 18:27:46 +0000

htop (2.1.0-2) unstable; urgency=medium

  * Add missing build-dependency on pkg-config

 -- Graham Inggs <ginggs@debian.org>  Mon, 05 Feb 2018 16:59:56 +0000

htop (2.1.0-1) unstable; urgency=medium

  [ Graham Inggs ]
  * New upstream release 2.1.0 (Closes: #889634)
  * Refresh debian/patches/601-openvz-new-ctid-vpid.patch
  * Update Vcs-* URIs for move to salsa.debian.org
  * Turn debhelper up to 11
  * Update debian/copyright
  * Remove trailing whitespace from debian/changelog
  * Use secure URI in Homepage and debian/watch
  * Bump Standards-Version to 4.1.3, no changes

  [ Helmut Grohne ]
  * Fix FTCBFS: Annotate python-minimal build dependency with :native.
    (Closes: #864672)

 -- Graham Inggs <ginggs@debian.org>  Mon, 05 Feb 2018 13:22:25 +0000

htop (2.0.2-1) unstable; urgency=medium

  * Update to upstream version 2.0.2
    - Use SOURCE_DATE_EPOCH instead of current year in copyright message
      to make build reproducible. (Closes: #822963)
    - Fix build to detect the Hurd as a Linux platform. (Closes: #821747)
    - Fix crash when setting CPU affinity and -fPIE. (LP: #1589549)
  * Bump Standards to 3.9.8 (no changes required)

 -- Daniel Lange <dl.ml1@usrlocal.de>  Mon, 25 Jul 2016 19:46:17 +0200

htop (2.0.1-1) unstable; urgency=medium

  * New upstream release 2.0.1. (Closes: #814401, #783893, #774930)
  * Clean up patches from 1.0.3 -> 2.0.1
  * Update compat -> 9, rules -> debhelper/9
  * Update debian/copyright to copyright-format/1.0 (Closes: #749850)
  * Add lsof as suggested (Closes: #817918)
  * Change maintainership jackyf -> DLange, jackyf and ginggs will
    co-maintain. Jackyf: Thank you very much for 8 years of htop
    maintenance!

 -- Daniel Lange <dl.ml1@usrlocal.de>  Thu, 14 Apr 2016 10:25:45 +0200

htop (1.0.3-1) unstable; urgency=low

  * New upstream release.
  * debian/watch:
    - Updated to use new download location (http://hisham.hm/htop/releases).
  * debian/patches:
    - 600-openvz-veid-on-kernels-gt-2.6.18.patch: refreshed.
    - 610-io-priority-only-on-linux.patch: ported to new upstream version.
    - 700-ltrace-support.patch: ported to new upstream version.
  * debian/control:
    - Bumped Standards-Version to 3.9.5, no changes needed.
    - Updated Homepage to 'http://hisham.hm/htop/'.

 -- Eugene V. Lyubimkin <jackyf@debian.org>  Thu, 01 May 2014 13:16:56 +0300

htop (1.0.2-3) unstable; urgency=low

  * debian/patches:
    - New patch 610-io-priority-only-on-linux.patch which should fix building
      on non-Linux kernels.

 -- Eugene V. Lyubimkin <jackyf@debian.org>  Sat, 05 Oct 2013 13:17:13 +0300

htop (1.0.2-2) unstable; urgency=low

  * Upload to unstable.

 -- Eugene V. Lyubimkin <jackyf@debian.org>  Sat, 29 Jun 2013 22:25:41 +0300

htop (1.0.2-1) experimental; urgency=low

  * New upstream release.
    - Ability to update process names on refreshes. (Closes: #475254)
    - More precise value of CPU%. (Closes: #668410)
    - Columns line up properly even when for more-than-5-digits process ids.
      (Closes: #687280)
  * debian/control:
    - Bumped Standards-Version to 3.9.4, no changes needed.
  * debian/patches:
    - Deleted 200-dont-delete-htoprc-if-symlink.patch, applied upstream.
    - 700-ltrace-support.patch: ported to new upstream version.

 -- Eugene V. Lyubimkin <jackyf@debian.org>  Sat, 01 Dec 2012 11:55:43 +0200

htop (1.0.1-4) unstable; urgency=low

  * debian/rules:
    - Retrieve and pass also CPPFLAGS and LDFLAGS. Patch by Simon Ruderich.
      (Closes: #685670)

 -- Eugene V. Lyubimkin <jackyf@debian.org>  Sat, 01 Sep 2012 14:29:59 +0300

htop (1.0.1-3) unstable; urgency=low

  * debian/rules:
    - Retrieve CFLAGS from dpkg-buildflags as dpkg-buildpackage stopped to
      export them. Thanks to Felix Geyer and Cyril Brulebois.
      (Closes: #685670)
  * debian/control:
    - Build-depend on dpkg (>= 1.16.1) for using dpkg-buildflags.

 -- Eugene V. Lyubimkin <jackyf@debian.org>  Mon, 27 Aug 2012 19:51:54 +0300

htop (1.0.1-2) unstable; urgency=low

  * debian/patches:
    - New patch 200-dont-delete-htoprc-if-symlink.patch from upstream VCS.
      (Closes: #662087)

 -- Eugene V. Lyubimkin <jackyf@debian.org>  Mon, 30 Jul 2012 19:12:16 +0300

htop (1.0.1-1) unstable; urgency=low

  * New upstream release.
    - Don't crash when trying to kill already finished process.
      (Closes: #652563, 657675)

 -- Eugene V. Lyubimkin <jackyf@debian.org>  Sun, 19 Feb 2012 10:24:46 +0200

htop (1.0-1) unstable; urgency=low

  * New upstream release.
    - Don't die after a terminal resize. (Closes: #464852)
    - Don't abort when a number of processes are removed from the list at
      once. (Closes: #541417)
    - Fixed the inconsistency in the documentation about "[" and "]" keys.
      (Closes: #610329)
  * debian/patches:
    - 100-fix-cgroups-crash.patch: deleted, upstream fixed it in a different
      way.
    - 600-openvz-veid-on-kernels-gt-2.6.18.patch: ported to new upstream
      version.
    - 700-ltrace-support.patch: ported to new upstream version.
    - 700-ltrace-support.patch: ltrace hotkey is 'L', not 'l'.
  * debian/control:
    - Add libhwloc-dev to Build-Depends on non-linux architectures.
    - Add libncursesw5-dev to Build-Depends, enabling unicode support.
  * debian/rules:
    - Enable native affinity functions on linux architectures and use hwloc
      library everywhere else.
  * debian/copyright:
    - Updated upstream copyright years.


linux/LinuxProcess.h
/* Do not edit this file. It was automatically generated. */

#ifndef HEADER_LinuxProcess
#define HEADER_LinuxProcess
/*
htop - LinuxProcess.h
(C) 2014 Hisham H. Muhammad
Released under the GNU GPL, see the COPYING file
in the source distribution for its full text.
*/


#define PROCESS_FLAG_LINUX_IOPRIO   0x0100
#define PROCESS_FLAG_LINUX_OPENVZ   0x0200
#define PROCESS_FLAG_LINUX_VSERVER  0x0400
#define PROCESS_FLAG_LINUX_CGROUP   0x0800
#define PROCESS_FLAG_LINUX_OOM      0x1000

typedef enum UnsupportedProcessFields {
   FLAGS = 9,
   ITREALVALUE = 20,
   VSIZE = 22,
   RSS = 23,
   RLIM = 24,
   STARTCODE = 25,
   ENDCODE = 26,
   STARTSTACK = 27,
   KSTKESP = 28,
   KSTKEIP = 29,
   SIGNAL = 30,
   BLOCKED = 31,
   SSIGIGNORE = 32,
   SIGCATCH = 33,
   WCHAN = 34,
   NSWAP = 35,
   CNSWAP = 36,
   EXIT_SIGNAL = 37,
} UnsupportedProcessField;

typedef enum LinuxProcessFields {
   CMINFLT = 11,
   CMAJFLT = 13,
   UTIME = 14,
   STIME = 15,
   CUTIME = 16,
   CSTIME = 17,
   M_SHARE = 41,
   M_TRS = 42,
   M_DRS = 43,
   M_LRS = 44,
   M_DT = 45,
   #ifdef HAVE_OPENVZ
   CTID = 100,
   VPID = 101,
   #endif
   #ifdef HAVE_VSERVER
   VXID = 102,
   #endif
   #ifdef HAVE_TASKSTATS
   RCHAR = 103,
   WCHAR = 104,
   SYSCR = 105,
   SYSCW = 106,
   RBYTES = 107,
   WBYTES = 108,
   CNCLWB = 109,
   IO_READ_RATE = 110,
   IO_WRITE_RATE = 111,
   IO_RATE = 112,
   #endif
   #ifdef HAVE_CGROUP
   CGROUP = 113,
   #endif
   OOM = 114,
   IO_PRIORITY = 115,
   #ifdef HAVE_DELAYACCT
   PERCENT_CPU_DELAY = 116,
   PERCENT_IO_DELAY = 117,
   PERCENT_SWAP_DELAY = 118,
   #endif
   LAST_PROCESSFIELD = 119,
} LinuxProcessField;

#include "IOPriority.h"

typedef struct LinuxProcess_ {
   Process super;
   IOPriority ioPriority;
   unsigned long int cminflt;
   unsigned long int cmajflt;
   unsigned long long int utime;
   unsigned long long int stime;
   unsigned long long int cutime;
   unsigned long long int cstime;
   long m_share;
   long m_trs;
   long m_drs;
   long m_lrs;
   long m_dt;
   #ifdef HAVE_TASKSTATS
   unsigned long long io_rchar;
   unsigned long long io_wchar;
   unsigned long long io_syscr;
   unsigned long long io_syscw;
   unsigned long long io_read_bytes;
   unsigned long long io_write_bytes;
   unsigned long long io_cancelled_write_bytes;
   unsigned long long io_rate_read_time;
   unsigned long long io_rate_write_time;   
   double io_rate_read_bps;
   double io_rate_write_bps;
   #endif
   #ifdef HAVE_OPENVZ
   char* ctid;
   unsigned int vpid;
   #endif
   #ifdef HAVE_VSERVER
   unsigned int vxid;
   #endif
   #ifdef HAVE_CGROUP
   char* cgroup;
   #endif
   unsigned int oom;
   char* ttyDevice;
   #ifdef HAVE_DELAYACCT
   unsigned long long int delay_read_time;
   unsigned long long cpu_delay_total;
   unsigned long long blkio_delay_total;
   unsigned long long swapin_delay_total;
   float cpu_delay_percent;
   float blkio_delay_percent;
   float swapin_delay_percent;
   #endif
} LinuxProcess;

#ifndef Process_isKernelThread
#define Process_isKernelThread(_process) (_process->pgrp == 0)
#endif

#ifndef Process_isUserlandThread
#define Process_isUserlandThread(_process) (_process->pid != _process->tgid)
#endif


extern ProcessFieldData Process_fields[];

extern ProcessPidColumn Process_pidColumns[];

extern ProcessClass LinuxProcess_class;

LinuxProcess* LinuxProcess_new(Settings* settings);

void Process_delete(Object* cast);

/*
[1] Note that before kernel 2.6.26 a process that has not asked for
an io priority formally uses "none" as scheduling class, but the
io scheduler will treat such processes as if it were in the best
effort class. The priority within the best effort class will  be
dynamically  derived  from  the  cpu  nice level of the process:
extern io_priority;
*/
#define LinuxProcess_effectiveIOPriority(p_) (IOPriority_class(p_->ioPriority) == IOPRIO_CLASS_NONE ? IOPriority_tuple(IOPRIO_CLASS_BE, (p_->super.nice + 20) / 5) : p_->ioPriority)

IOPriority LinuxProcess_updateIOPriority(LinuxProcess* this);

bool LinuxProcess_setIOPriority(LinuxProcess* this, IOPriority ioprio);

#ifdef HAVE_DELAYACCT
void LinuxProcess_printDelay(float delay_percent, char* buffer, int n);
#endif

void LinuxProcess_writeField(Process* this, RichString* str, ProcessField field);

long LinuxProcess_compare(const void* v1, const void* v2);

bool Process_isThread(Process* this);


#endif
